### ISSUE TITLE
Put all oe_structure in separate views

### ISSUE THREAD

**Issue by Yajo** [odoo/odoo#15437]
Impacted versions: 8, 9, 10, 11

Steps to reproduce:

1. Install website_sale.
2. Go to cart (buy something...).
3. Edit the page.
4. Insert something in an editable area.
5. Go to backend in debug mode.
5. Go to *Settings > Technical > Sequences & Identifiers > External identifiers*.
6. Search for `website_sale.cart` XMLID.

Current behavior: View gets marked as `noupdate=True`. This means that, if a patch lands to the cart and you update the addon, **you will get no patch**. Also means that when migrating from older to newer versions, you will get no updates there. Then you will have to go view by view checking out what changes did you make, saving that somewhere else, setting is at `noupdate=False`, updating the addon, restoring your changes, and waiting to do the same procedure for next version. 😞 

Expected behavior: Editable parts in views should be separate views.

Proposed solution:

For instance, having this view:

```
<template id="some_view">
  <!-- Header editable area -->
  <div class="oe_structure oe_empty"/>

  <!-- Parts that need some logic and should get updates -->
  <!-- ... -->

  <!-- Footer editable area -->
  <div class="oe_structure oe_empty"/>
</template>
```

It should be rewritten like this:

```
<template id="some_view">
  <!-- Header editable area -->
  <t t-call="custom_01"/>

  <!-- Parts that need some logic and should get updates -->
  <!-- ... -->

  <!-- Footer editable area -->
  <t t-call="custom_02"/>
</template>

<!-- User-writable dummy templates -->
<template id="custom_01">
  <div class="oe_structure oe_empty"/>
</template>
<template id="custom_02">
  <div class="oe_structure oe_empty"/>
</template>
```

This way you will not need to choose between having a customized website and having a maintainable one.

If you still manually update the non-editable parts (which can be done, not sure why), then it's OK for me that you get no updates. It's what you pay. But if you just want to add some custom photo, message, or whatever in a user-editable section, it should cost nothing, maintainability-wise. Obvoiusly, all these `custom_XX` views should be acknowledged to have no logic at all, and just be placeholders for user overridings.

I'm OK to patch this myself in PR for master, but it's a big job and I want to know opinions from upstream -- and if it would get approval -- before starting it.

cc @JKE-be @Tecnativa

**Comment by simahawk** [comment]
:+1: is something I always do in my custom views to prevent this issue.

**Comment by Yenthe666** [comment]
@JKE-be this one has been lying around for quite a while. Can you give feedback and if this would be accepted? Would like to get this handled before V11.

**Comment by odony** [comment]
I think it could be a good improvement, but rather than having a bunch of empty views and as many t-calls, I'd rather have 1 extra view for each page where we need to add editable zones. The extra view would be a sort of "customization layer" on top of the main one.
That should work for the same purpose, with less verbosity, shouldn't it?

i.e something like:
```xml
<template id="main_view">
  <div class="o_custom_anchor">
     <!-- Parts that need some logic and should get updates -->
     <!-- ... -->
  </div>
</template>

<template id="main_view_customizations" inherit_id="main_view">
  <xpath expr="//div[hasclass('o_custom_anchor')]" position="before">
     <!-- Header editable area -->
     <div class="oe_structure oe_empty"/>
  </xpath>
  <xpath expr="//div[hasclass('o_custom_anchor')]" position="after">
     <!-- Footer editable area -->
     <div class="oe_structure oe_empty"/>
  </xpath>
</template>
```

**Comment by Yajo** [comment]
I think the `t-call` approach is better. Example:

```xml
<template id="module.layout" page="True">
  <head>
    <t t-call="module.header"/>
  </head>
  <body>
    <t t-call="module.body"/>
  </body>
  <footer>
    <t t-call="module.footer"/>
  </footer>
</template>

<template id="header">
  <div class="oe_structure oe_empty"/>
</template>
<template id="body">
  <div class="oe_structure oe_empty"/>
</template>
<template id="footer">
  <div class="oe_structure oe_empty"/>
</template>
```

Now I have a different addon that removes the footer for any reason:

```xml
<template id="other.layout" inherit_id="module.layout">
  <xpath expr="//footer" position="replace"/>
</template>
```

Advantages: 

- Module installs and footer gets removed.
- If I want to put the footer *in a different view*, it's perfectly possible.
- The developer that reads the `module.layout` view immediately knows that it uses the other views when he sees the `t-call`s.

---

The same using replaces. Base:

```xml
<template id="module.layout" page="True">
  <head class="myhead"/>
  <body class="mybody"/>
  <footer class="myfooter"/>
</template>

<template id="header">
  <xpath expr="//+[hasClass('myhead')">
    <div class="oe_structure oe_empty"/>
  </xpath>
</template>
<template id="body">
  <xpath expr="//+[hasClass('mybody')">
    <div class="oe_structure oe_empty"/>
  </xpath>
</template>
<template id="footer">
  <xpath expr="//+[hasClass('myfooter')">
    <div class="oe_structure oe_empty"/>
  </xpath>
</template>
```

Footer removal:

```xml
<template id="other.layout" inherit_id="module.layout">
  <xpath expr="//footer" position="replace"/>
</template>
```

Problems:

- Module `other` would crash at install because `module.footer` looks for a replacement that it cannot find.
- It is way more verbose.
- There is no way to relocate the partial to a different view.
- The developer that reads `module.layout` has no idea it is being replaced by the other views unless he explicitly goes there.

---

So, I feel the `t-call` approach is way better.

**Comment by odony** [comment]
The point of having a single view to hold the customizable parts is to keep things simple. This includes keeping related customizations together, after being edited.
Your advantages sound like over-engineering to work around unlikely problems. Allowing partial relocation is not something we're trying to achieve. Modules breaking other modules by brutally removing DOM elements will happen no matter what, it's a bad practice.
Developers reading the XML file will see the structure in either case, the views will be next to each other. And it's a common pattern for "secondary" parts of a page. Often combined with `customize_show` to make it an explicit toggle.

**Comment by Yajo** [comment]
Well, it's not about over-engineering. Fixing this means a massive refactor of an under-engineered feature, and it's very unlikely that you want to massively refactor yet again if you (or others in the community) hit any of my points in the future.

The amount of work is going to be mostly the same in both cases, with the only difference that we will have more dummy views in the future. Is that *really* a problem? If you refactor as you exposed, chances are that some day you shoot yourself in the toe, while my idea is future-proof, and gives those extra possibilities that you may or not want to use, but you actually would have them.

Besides, imagine you have the main view with 3 placeholders in Odoo 11, and in Odoo 12 you need to add a 4th. Your update would forbid updated users to use the 4th, while mine would let them use it too without any hassle.

The opposite case: main view with 3 placeholders in v11, and you see that one of them was useless so you leave only 2 for v12. Your update would break the system while mine would just lose the now useless view.

Given the amount of work is the same, I'd honestly go for the path that means more features.

**Comment by odony** [comment]
@simahawk that's besides the point ;-)

**Comment by Yajo** [comment]
Yeah, a little ugly, but can you think of another solution? 😕 It would be less ugly than getting broken websites on every version update (really friend... *every single migration*).

**Comment by pedrobaeza** [comment]
Well, even if none of both solutions are the optimal, we should implement at least one for avoiding the current problems. I prefer @Yajo's solution because it allows to grow easily and the minimum overhead of having n templates when you want n placeholders is minimum, but I will like the other the same if we get this to reach v11. @odony, do you want us to code any of them?

**Comment by pedrobaeza** [comment]
@qsm-odoo Another one that involves you! Any plans to support this or to think about a more automatic mechanism? We can make a workshop about this if you want.

**Comment by pedrobaeza** [comment]
@JKE-be @qsm-odoo this is another of the rough edges we have found in website. Do you have any plan to change this? Maybe are you coming to the code sprint in Sevilla and can discuss this together?

**Comment by pedrobaeza** [comment]
@JKE-be can you update the status of this after the big merge of yesterday?

**Comment by qsm-odoo** [comment]
Merged a third solution for 12.0: https://github.com/odoo/odoo/commit/dd139948f0049b3676b859495ba9b2d43436f9c1
